network/p2p/acp118/aggregator.go

```diff
@@ -0,0 +1,260 @@
+// the message. Caller is responsible for providing a well-formed canonical
+// validator set corresponding to the signer bitset in the message.
+func (s *SignatureAggregator) AggregateSignatures(
+	ctx context.Context,
+	message *warp.Message,
+	justification []byte,
+	validators []*warp.Validator,
+	quorumNum uint64,
+	quorumDen uint64,
+) (
+	_ *warp.Message,
+	aggregatedStake *big.Int,
+	totalStake *big.Int,
+	finished bool,
+	_ error,
+) {
+	request := &sdk.SignatureRequest{
+		Message:       message.UnsignedMessage.Bytes(),
+		Justification: justification,
+	}
+
+	requestBytes, err := proto.Marshal(request)
+	if err != nil {
+		return nil, nil, nil, false, fmt.Errorf("failed to marshal signature request: %w", err)
+	}
+
+	publicKeysToValidators := make(map[*bls.PublicKey]indexedValidator)
+	nodeIDsToPublicKeys := make(map[ids.NodeID]*bls.PublicKey)
+	// TODO expose concrete type to avoid type casting
+	bitSetSignature, ok := message.Signature.(*warp.BitSetSignature)
+	if !ok {
+		return nil, nil, nil, false, errors.New("invalid warp signature type")
+	}
+
+	signerBitSet := set.BitsFromBytes(bitSetSignature.Signers)
+
+	nonSigners := make([]*bls.PublicKey, 0, len(validators))
+	aggregatedStakeWeight := new(big.Int)
+	totalStakeWeight := new(big.Int)
+	numRequests := 0
+	for i, v := range validators {
+		totalStakeWeight.Add(totalStakeWeight, new(big.Int).SetUint64(v.Weight))
+
+		// Only try to aggregate signatures from validators that are not already in
+		// the signer bit set
+		if signerBitSet.Contains(i) {
+			aggregatedStakeWeight.Add(aggregatedStakeWeight, new(big.Int).SetUint64(v.Weight))
+			continue
+		}
+
+		publicKeysToValidators[v.PublicKey] = indexedValidator{
+			Index:     i,
+			Validator: v,
+		}
+
+		for _, nodeID := range v.NodeIDs {
+			numRequests += 1
+			nodeIDsToPublicKeys[nodeID] = v.PublicKey
+		}
+
+		nonSigners = append(nonSigners, v.PublicKey)
+	}
+
+	// Account for requested signatures + the signature that was provided
+	signatures := make([]*bls.Signature, 0, len(nonSigners)+1)
+	if bitSetSignature.Signature != [bls.SignatureLen]byte{} {
+		blsSignature, err := bls.SignatureFromBytes(bitSetSignature.Signature[:])
+		if err != nil {
+			return nil, nil, nil, false, fmt.Errorf("failed to parse bls signature: %w", err)
+		}
+		signatures = append(signatures, blsSignature)
+	}
+
+	results := make(chan result)
+	handler := responseHandler{
+		message:                message,
+		publicKeysToValidators: publicKeysToValidators,
+		nodeIDsToPublicKeys:    nodeIDsToPublicKeys,
+		results:                results,
+	}
+
+	for _, pk := range nonSigners {
+		validator := publicKeysToValidators[pk]
+		for _, nodeID := range validator.NodeIDs {
+			if err := s.client.AppRequest(ctx, set.Of(nodeID), requestBytes, handler.HandleResponse); err != nil {
+				return nil, nil, nil, false, fmt.Errorf("failed to send aggregation request: %w", err)
+			}
+		}
+	}
```

Comment:
Should we eliminate the inner for loop and call `AppRequest` on all of the specified nodeIDs at once since `AppRequest` takes in a set?

Reply:
Sure